server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -9,30 +9,54 @@
 package org.elasticsearch.search.fetch.subphase;
 
 import org.apache.lucene.index.LeafReaderContext;
-import org.elasticsearch.common.document.DocumentField;
-import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.index.mapper.IgnoredFieldMapper;
+import org.elasticsearch.index.mapper.LegacyTypeFieldMapper;
+import org.elasticsearch.index.mapper.RoutingFieldMapper;
 import org.elasticsearch.search.fetch.FetchContext;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
 import org.elasticsearch.search.fetch.StoredFieldsSpec;
 
 import java.io.IOException;
-import java.util.Map;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
 
 /**
  * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
  * retrieves the field values through the relevant {@link org.elasticsearch.index.mapper.ValueFetcher}
  * and returns them as document fields.
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
+    private static final List<FieldAndFormat> METADATA_FIELDS = List.of(
+        new FieldAndFormat(IgnoredFieldMapper.NAME, null),
+        new FieldAndFormat(RoutingFieldMapper.NAME, null),
+        new FieldAndFormat(LegacyTypeFieldMapper.NAME, null),
+        new FieldAndFormat("_size", null)
+    );
+
+    public static boolean isMetadataField(final String field) {
+        return METADATA_FIELDS.stream().map(fieldAndFormat -> fieldAndFormat.field).anyMatch(fieldName -> fieldName.equals(field));
+    }
```

Comment:
I think that this is problematic, because you have a separate definition of metadata fields within fetch fields. Plugins (like `SizeMapperPlugin`) can provide their own extensions of custom metadata fields. When they do so, we should treat those metadata fields like any other. We certainly should not have the knowledge of which metadata fields plugins may provide. Ideally, we would simply rely on `SearchExecutionContext#isMetadataField` here, like `StoredFieldsPhase` used to do. Additionally , you need to skip the `_source` field exactly like `StoredFieldsPhase` used to do.